libraries/Microsoft.Bot.Builder/TokenResolver.cs

```diff
@@ -37,15 +50,16 @@ public static void CheckForOAuthCards(BotFrameworkAdapter adapter, ILogger logge
                         throw new InvalidOperationException("The OAuthPrompt's ConnectionName property is missing a value.");
                     }
 
-                    // Poll as a background task and add to the list (we don't call await here, we await all the calls together later).
                     pollTokenTasks.Add(PollForTokenAsync(adapter, logger, turnContext, oauthCard.ConnectionName, cancellationToken));
                 }
             }
 
             if (pollTokenTasks.Any())
             {
-                // Wait for all the poll operations to complete.
-                Task.WaitAll(pollTokenTasks.ToArray());
+                // Run the poll operations in the background.
+                // On retrieving a token from the token service the TokenResolver creates an Activity to route the token to the bot to continue the conversation.
+                // If these Tasks are awaited and the user doesn't complete the login flow, the bot may timeout in sending its response to the channel which can cause the streaming connection to disconnect.
+                Task.WhenAll(pollTokenTasks.ToArray());
```

Comment:
Since we are at it, it would be great to spend a bit of time making sure that any errors or timeouts in these polling tasks, there is a way for developers to detect them. Since these are ran in the background, the output status code is likely not the best avenue. However we do have a logger instance available and the ability to detect timeouts and failures in these tasks. If that seems of value, seems we could do a couple of things. 1) Catch those timeout and error scenarios. Code like the following could be in place

```csharp
Task.WhenAll(pollTokenTasks.ToArray()).ContinueWith((task) =>
{
    if (task.IsCanceled)
    {
        // handle timeout logging
    }
    else if (task.IsFaulted)
    {
        // handle error, exceptions can be found in task.Exception.
    }
});
```

Ideally, this would be accompained by tests for timeout and fault scenario but not sure off hand how much ability we have to inject / unit test that behavior.

Reply:
I agree with surfacing failures to the developer and I think it's handled in the `TokenResolver.PollForTokenAsync()` helper ([code](https://github.com/microsoft/botbuilder-dotnet/blob/2a71e744497469e5e7003acf1e0c43f149806df1/libraries/Microsoft.Bot.Builder/TokenResolver.cs#L114-L136)). When I switched [`WaitAll()` to `WhenAll()`](https://github.com/microsoft/botbuilder-dotnet/blob/2a71e744497469e5e7003acf1e0c43f149806df1/libraries/Microsoft.Bot.Builder/TokenResolver.cs#L62) I thought about adding the [`ContinueWith()` logic from 4.9](https://github.com/microsoft/botbuilder-dotnet/blob/4.9/libraries/Microsoft.Bot.Builder/TokenResolver.cs#L42-L49). However, in R10 the `PollForTokenAsync()` logic was updated so that all exceptions are caught, so I didn't think it was necessary to restore the `ContinueWith()`

However, I haven't tested this path in an E2E scenario, so I'm not sure if the code works according to design 🙁

@Virtual-Josh was the last person to update the tests for this class, can you confirm if we have unit test coverage for surfacing errors?

___

Regarding unit testing this class, we can stand up `Moqs` to send back responses for 401s, the number of 404s, 5xxs to test both the number of polling operations and non-happy paths

Reply:
Agreed on the ContinueWith, not needed because of the new try-catch.

Regarding testing, adding negative tests to this code area would be great. Not mandatory for this PR but something I'd love to see in R11 as part of our investment to make this more solid.